File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        Map<String, Integer> wantedCount = new HashMap<>();
        for (int i = 0; i < want.length; i++) {
            wantedCount.put(want[i], number[i]);
        }

        Map<String, Integer> discountCount = new HashMap<>();
        
        int answer = 0;
        int right = 0;
        for (int left = 0; left <= discount.length - 10; left++) {

            while(right - left < 10) {
                discountCount.put(discount[right], discountCount.getOrDefault(discount[right], 0) + 1);
                right++;
            }
            if (discountCount.equals(wantedCount)) {
                answer++;
            }

            int cnt = discountCount.get(discount[left]) - 1;
            if (cnt <= 0) {
                discountCount.remove(discount[left]);
            } else {
                discountCount.put(discount[left], cnt);
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
using namespace std;

map<string, int> m;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    int days = 0;
    
    for(int i = 0; i < want.size(); ++i) {
        m.insert(make_pair(want[i], i));
    }
    
    for(int i = 0; i < discount.size(); ++i) {
        if (m.find(discount[i]) != m.end())
            --number[m[discount[i]]];
        
        if (i - 10 >= 0) {
            if (m.find(discount[i-10]) != m.end())
                ++number[m[discount[i-10]]];
        }
        
        bool flag = true;
        for(int j = 0; j < number.size() && flag; ++j) {
            if (number[j] != 0)
                flag = false;
        }
        
        if (flag) {
            cout << i + 1 << endl;
            ++answer;
        }
        
    }
    
    return answer;
}